cmd/dagger/call.go

```diff
@@ -1,253 +1,40 @@
-// - [ ] Create multiple "verb" commands for calling functions with different intents.
-// - [ ] Add `dagger shell` as a high priority for its debugging usefulness.
-// - [ ] Add a --help flag to list available functions.
-// - [ ] Maybe a list/functions command to list all functions and compatible verbs that can be used to call them.
-// - [ ] Maybe compose a query instead of using `Function.call`?
-// - [ ] Use `dagger run` instead of `dagger call` for generic execution.
-//       Detect if running in the context of a module for backwards compatibility.
-// - [ ] Improve progrock usage.
-// - [ ] Support non-progrock (e.g., --progress plain) output.
-// - [ ] Fix Go SDK's listing bug.
-// - [ ] Add support for chaining function calls.
-
-// outputPath is the host directory to export assets to.
-// var outputPath string
-
-var callCmd = &cobra.Command{
-	Use:    "call [function]",
-	Long:   `Call a module's function.`,
-	RunE:   loadModCmdWrapper(RunFunction, ""),
-	Hidden: true,
-}
-
-func init() {
-	rootCmd.AddCommand(callCmd)
-
-	// TODO: Use `dagger download` verb.
-	// callCmd.PersistentFlags().StringVarP(&outputPath, "output", "o", "", "If the function returns a file or directory, it will be written to this path. If --output is not specified, the file or directory will be written to the module's root directory when using a module loaded from a local dir.")
-	callCmd.PersistentFlags().BoolVar(&focus, "focus", true, "Only show output for focused commands.")
-
-	callCmd.AddCommand(
-		&cobra.Command{
-			Use:          "list",
-			Long:         `List your module's functions.`,
-			SilenceUsage: true,
-			RunE:         loadModCmdWrapper(ListFunctions, ""),
-			Hidden:       true,
-		},
-	)
-}
-
-func ListFunctions(ctx context.Context, engineClient *client.Client, mod *dagger.Module, cmd *cobra.Command, cmdArgs []string) (err error) {
-	if mod == nil {
-		return fmt.Errorf("no module specified and no default module found in current directory")
-	}
-
-	rec := progrock.FromContext(ctx)
-	vtx := rec.Vertex("cmd-list-functions", "list functions", progrock.Focused())
-	defer func() { vtx.Done(err) }()
-
-	modName, err := mod.Name(ctx)
-	if err != nil {
-		return fmt.Errorf("failed to get module name: %w", err)
-	}
-
-	modObjs, err := mod.Objects(ctx)
-	if err != nil {
-		return fmt.Errorf("failed to get module objects: %w", err)
-	}
-
-	var modFuncs []dagger.Function
-	for _, def := range modObjs {
-		objDef := def.AsObject()
-		objName, err := objDef.Name(ctx)
-		if err != nil {
-			return fmt.Errorf("failed to get object name: %w", err)
-		}
-
-		// TODO: Only list top level functions for now (i.e., type name matches module name).
-		if strcase.ToLowerCamel(objName) != strcase.ToLowerCamel(modName) {
-			continue
+var callCmd = &FuncCommand{
+	Name:  "call",
+	Short: "Call a module's function.",
+	OnSelectObject: func(c *callContext, name string) (*modTypeDef, error) {
+		switch name {
+		case Container:
```

Comment:
Not that this would make any sense, but if someone tried to call a Function that returned Secret, does that cause `dagger call` to error out (probably most logical behavior)? Haven't seen we handle that, but may be missing it.

Reply:
On `call` being generic, I'm not sure what the behavior should be for "unknown" (or not specifically handled) returns.

Right now the default is just `cmd.Print("%+v\n", result)` (in `OnResult`), so for a `Secret` it would show it's `SecretID`.

I’m leaning to just print by default for now and bikeshed later, unless there’s some specific we should not print.